extensions-core/hdfs-storage/src/main/java/io/druid/storage/hdfs/HdfsDataSegmentPusher.java

```diff
@@ -88,24 +89,43 @@ public DataSegment push(File inDir, DataSegment segment) throws IOException
         storageDir
     );
 
-    Path outFile = new Path(String.format("%s/%s/index.zip", config.getStorageDirectory(), storageDir));
-    FileSystem fs = outFile.getFileSystem(hadoopConfig);
+    Path tmpFile = new Path(String.format(
+        "%s/%s/index.zip",
+        config.getStorageDirectory(),
+        UUIDUtils.generateUuid()
+    ));
+    FileSystem fs = tmpFile.getFileSystem(hadoopConfig);
 
-    fs.mkdirs(outFile.getParent());
-    log.info("Compressing files from[%s] to [%s]", inDir, outFile);
+    fs.mkdirs(tmpFile.getParent());
+    log.info("Compressing files from[%s] to [%s]", inDir, tmpFile);
 
     final long size;
-    try (FSDataOutputStream out = fs.create(outFile)) {
+    final DataSegment dataSegment;
+    try (FSDataOutputStream out = fs.create(tmpFile)) {
       size = CompressionUtils.zip(inDir, out);
+      dataSegment = createDescriptorFile(
+          segment.withLoadSpec(makeLoadSpec(tmpFile))
+                 .withSize(size)
+                 .withBinaryVersion(SegmentUtils.getVersionFromDir(inDir)),
+          tmpFile.getParent(),
+          fs
+      );
+      Path outDir = new Path(String.format("%s/%s", config.getStorageDirectory(), storageDir));
+      if (!fs.rename(tmpFile.getParent(), outDir)) {
+        if(!fs.delete(tmpFile.getParent(), true)){
+          log.error("Failed to delete temp directory");
+        }
+        if(fs.exists(outDir)){
```

Comment:
@du00cs: I think ideally, if one of the replica task lost the race, the other ones should not throw the exception here, a log message here instead of an exception if the outDir exists should be fine. Are you able to quickly fix this up ?